File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> answer(n);
        int prefixProduct = 1;
        for (int i = 0; i<n; i++){
            answer[i] = prefixProduct;
            prefixProduct *= nums[i];
        }

        int postfixProduct = 1;
        for (int i = n-1; i>=0; --i){
            answer[i] *= postfixProduct;
            postfixProduct *= nums[i];
        }
        return answer;
    }
};

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        pro = 1
        prod_pre = [1]*len(nums)
        prod_post = [1]*len(nums)
        for i in range(len(nums)):
            prod_pre[i] = nums[i] * pro
            pro = prod_pre[i]
        pro = 1
        for i in range(len(nums)-1, -1, -1):
            prod_post[i] =  pro * nums[i]
            pro = prod_post[i]

        answer=[]
        for i in range(len(nums)):
            if i == 0:
                answer.append(prod_post[i+1])
            elif i == len(nums)-1:
                answer.append(prod_pre[i-1])
            else:
                answer.append(prod_post[i+1]*prod_pre[i-1])
        
        return answer